Support ERC4337 UserOperation Signature Requests

KPrasch opened the pull request:
**Type of PR:**
Feature

**Required reviews:** 
2

**What this does:**

This PR introduces first-class support for multi-EIP signature types within the NuCypher signing protocol. Specifically:
- Adds SignatureRequest and SignatureResponse data structures for EIP-712 and EIP-191 signature types
- Deprecates ThresholdSignatureRequest/Response in favor of unified, EIP-aware classes
- Renames network/decryption.py to network/concurrency.py
- Implements `sign_message_eip191` and `sign_message_eip712` in both TransactingPower and all signer classes
- Introduces `nucypher.utilities.erc4337_utils` module for ERC-4337-compatible PackedUserOperation handling, including:
    - EIP-712 encoding
    - OZ and Infinitism calldata packing
    - Hashing and signing support
    - Helpers for ETH and ERC20 operations
- Adds extensive test coverage in test_signing_ritual.py and supporting test utilities
- Adds support for signing requests through REST endpoints with proper signature type discrimination

KPrasch committed:
exposes message hash from signed messages

KPrasch committed:
Updates Signature request protocol objects to handle multiple eips.

KPrasch committed:
exposing eip 712 encoding on power internals

KPrasch committed:
Introduces UserOp EIP-712 structures.

KPrasch committed:
preperations for userop 712 compatibility

KPrasch committed:
introduces er4337_utils.py module (test utility... for now)

KPrasch committed:
renames decryption.py -> concurrency.py

KPrasch committed:
implements eip-712 / eip-191 signature request+response parameterization

KPrasch committed:
include user operation packing compatible with both the raw 4337 standard as well as OZ modifications.

KPrasch committed:
fixes hashing methods of PackedUserOperation

KPrasch committed:
extends acceptance testing of user operation signing

KPrasch committed:
implements PackedUserOperation.sign

KPrasch committed:
use the domain string 'ERC4337'

KPrasch committed:
implements signature request type for userop.

derekpierre committed:
Post-rebase cleanup.

Co-authored-by: Kieran Prasch <kieranprasch@gmail.com>

derekpierre committed:
Code cleanup.
Signing returns tuple of HexBytes.
Skip erc4337 tests.

theref commented on nucypher/blockchain/eth/actors.py:
```diff
@@ -1015,30 +1023,40 @@ def handle_threshold_signing_request(
+            # TODO: userop stuff (validate?)
+            message_hash, signature = userop.sign(
+                transacting_power=self.transacting_power,
+                entrypoint=EntryPointContracts.ENTRYPOINT_V08,
```
this will need to be parameterized `v6 | v7 | v8`

derekpierre replied:
✔️

theref commented on nucypher/utilities/erc4337_utils.py:
```diff
@@ -0,0 +1,265 @@
+    return HexBytes(b"")
+
+
+class EntryPointContracts:
```
🔥

theref reviewed (dismissed):
not sure if you came across this, might be useful https://docs.apeworx.io/eip712/stable/

KPrasch committed:
rrenames ThresholdSigningClient -> SigningRequestClient

KPrasch committed:
expands unit testing for erc4337, unskip tests post-rebase.

KPrasch committed:
Disable generic message signing in favor of purpose-built UserOperationSigning flow

derekpierre committed:
Split and restructure of UserOperation and PackedUserOperation (created from UserOperation).

derekpierre committed:
Modify hierarchy of BaseSignatureRequest and sub-classes, usagei, and serialization.
Limiting the logic a bit since these classes will move into `nucypher-core`.

derekpierre committed:
Use a test contract to verify message hash calculations for PackedUserOperation.
Create associated test.

derekpierre committed:
Use oz 5.3.0 to further reduce additional code in EntryPoint test contract.
Change name of 5.3.0 package to prevent naming clashes with oz 5.0.0 contracts used via nucypher-contracts.

derekpierre committed:
Maintain signature request objects as pure data objects, so move sign() function out of object and into its own function.
Update EIPSignatureRequest to properly parse its data from bytes, and add it as a type to be deserialized.

derekpierre committed:
Update test to include test for EIP191 signing - just for testing purposes at the moment.

derekpierre committed:
Use AAVersion instead of EntryPointVersion.
Add version enum value for MetaMask Delegation Toolkit - this is a workaround for now.
Add/update tests.

derekpierre committed:
Add test for validating our packing of gas limits and gas fees.

derekpierre committed:
Re-instate various paymaster values and test the packing of them appropriately.

Co-authored-by: Kieran Prasch <kieranprasch@gmail.com>

derekpierre committed:
Rename test file to be more accurate.

derekpierre committed:
Rename already existing newsfragment since 3594 was closed in favour of 3599.

derekpierre committed:
Appease linter.

derekpierre committed:
Fix testnet signing script to use updated attribute hash instead of message_hash.

derekpierre committed:
Be more lenient with missing fields in serialized bytes.

derekpierre committed:
Make adjustments for special MDT case based on latest investigation.

derekpierre committed:
Testing out theory of having PackedUserOperation be directly provided.

derekpierre committed:
The verifying contract used for the domain value should be the sender contract i.e. the multisig contract itself.

derekpierre committed:
Update EntryPoint test contract to update MDT calculation.

derekpierre commented:
Good milestone for merging into epic - now that the MDT test script is working.

@KPrasch I could have improved codecov a bit, but ran out of time - perhaps we deal with that in a separate PR, wdyt?
